### PR TITLE
FIX: delete solution with post

### PR DESCRIPTION
**Pull request by ZogStriP**
Ensures we remove the solution when the post marked as the solution is deleted.

DEV: Added `IS_ACCEPTED_ASNWER_CUSTOM_FIELD` constant.
DEV: Refactored the `PostSerializer` for better readability.
PERF: Improved the `TopicViewSerializer`'s performance by looking up the `accepted_answer_post_info` from the stream first.

Internal ref. dev/112251

### CODE REVIEW

File: plugin.rb [hunk 1]
```diff
@@ -77,6 +77,7 @@ class Engine < ::Rails::Engine
     AUTO_CLOSE_TOPIC_TIMER_CUSTOM_FIELD = "solved_auto_close_topic_timer_id"
     ACCEPTED_ANSWER_POST_ID_CUSTOM_FIELD = "accepted_answer_post_id"
     ENABLE_ACCEPTED_ANSWERS_CUSTOM_FIELD = "enable_accepted_answers"
+    IS_ACCEPTED_ASNWER_CUSTOM_FIELD = "is_accepted_answer"
```
Comment (ZogStriP) [inline plugin.rb:80]: Added this constant so we don't keep referencing the `"is_accepted_answer"` string everywhere.

File: plugin.rb [hunk 2]
```diff
@@ -450,29 +458,27 @@ def accepted_answer
...
-          .where(id: accepted_answer_post_id, topic_id: object.topic.id)
-          .joins(:user)
-          .pluck("post_number", "username", "cooked", "name")
-          .first
-
-      if postInfo
-        postInfo[2] = if SiteSetting.solved_quote_length > 0
-          PrettyText.excerpt(postInfo[2], SiteSetting.solved_quote_length, keep_emoji_images: true)
+      post_info =
+        if post = object.posts.find { |p| p.post_number == accepted_answer_post_id }
```
Comment (ZogStriP) [inline plugin.rb:462]: This is the "performance" change which looks up the stream for the post that is marked as the solution before doing a request to the database.

File: plugin.rb [hunk 3]
```diff
@@ -450,29 +458,27 @@ def accepted_answer
...
 
-        postInfo[3] = (
-          if SiteSetting.enable_names && SiteSetting.display_name_on_posts
-            postInfo[3]
-          else
-            nil
-          end
-        )
-        postInfo
+        post_info[3] = nil if !SiteSetting.enable_names || !SiteSetting.display_name_on_posts
```
Comment (ZogStriP) [inline plugin.rb:479]: Feels like it's clearer to "just" set to `nil` when names are disabled.

File: spec/integration/solved_spec.rb [hunk 4]
```diff
@@ -268,6 +268,23 @@
       expect(p1.custom_fields["is_accepted_answer"]).to eq("true")
     end
 
+    it "removes the solution when the post is deleted" do
```
Comment (ZogStriP) [inline spec/integration/solved_spec.rb:271]: The test for the gist of the PR

File: plugin.rb [hunk 5]
```diff
@@ -543,32 +549,28 @@ def can_accept_answer?(topic, post)
...
     end
 
     def topic_accepted_answer
-      if topic = (topic_view && topic_view.topic) || object.topic
-        topic.custom_fields[::DiscourseSolved::ACCEPTED_ANSWER_POST_ID_CUSTOM_FIELD].present?
-      end
+      topic&.custom_fields[::DiscourseSolved::ACCEPTED_ANSWER_POST_ID_CUSTOM_FIELD].present?
+    end
+
+    def topic
```
Comment (ZogStriP) [inline plugin.rb:572]: Tiny refactor for improved readability.

File: plugin.rb (hunk 1 above)
Comment (pmusaraj) [inline plugin.rb:80]:
```suggestion
    IS_ACCEPTED_ANSWER_CUSTOM_FIELD = "is_accepted_answer"
```

little typo (I would imagine some tests will fail?)

File: spec/integration/solved_spec.rb [hunk 6]
```diff
@@ -181,7 +181,7 @@
 
       expect(topic.public_topic_timer.status_type).to eq(TopicTimer.types[:silent_close])
 
-      expect(topic.custom_fields[DiscourseSolved::AUTO_CLOSE_TOPIC_TIMER_CUSTOM_FIELD].to_i).to eq(
+      expect(topic.custom_fields["solved_auto_close_topic_timer_id"].to_i).to eq(
```
Comment (pmusaraj) [inline spec/integration/solved_spec.rb:184]: Curious, why not keep the constant? For readability?

**Review by pmusaraj — APPROVED** [review]

File: spec/integration/solved_spec.rb (hunk 6 above)
Comment (ZogStriP) [reply spec/integration/solved_spec.rb:184]: Felt like in specs it was shorter to use the string instead of the constant.

File: plugin.rb (hunk 1 above)
Comment (ZogStriP) [reply plugin.rb:80]: No test failed because I typoed the constant everywhere ;)

**Pull request merged**